webhooks/pod_webhook.go

```diff
@@ -230,6 +257,146 @@ func (m *PodMutator) injectSidecar(
+	d := appsV1.Deployment{}
+	err := m.Client.Get(ctx, client.ObjectKey{Name: kubeProxyDeploymentName, Namespace: currentNamespace}, &d)
+	if err != nil {
+		if errors.IsNotFound(err) {
+			// does not exist, is not ready, no error
+			return false, false, nil
+		}
+		// does not exist, is not ready, is in error
+		return false, false, err
+	}
+	if d.Status.ReadyReplicas == 0 {
+		// exists, not ready, no error
+		return true, false, nil
+	}
+	// exists, at least one replica ready, no error
+	return true, true, nil
+}
+
+func (m *PodMutator) deployKubeProxy(ctx context.Context) error {
+	m.Log.Info("deploying the kube-flagd-proxy")
+	if err := m.Client.Create(ctx, newFlagdKubeProxyManifest()); err != nil && !errors.IsAlreadyExists(err) {
+		return err
+	}
+	m.Log.Info("deploying the kube-flagd-proxy service")
+	if err := m.Client.Create(ctx, newFlagdKubeProxyServiceManifest()); err != nil && !errors.IsAlreadyExists(err) {
+		return err
+	}
+	return nil
+}
+
+func newFlagdKubeProxyServiceManifest() *corev1.Service {
+	return &corev1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      kubeProxyServiceName,
+			Namespace: currentNamespace,
+		},
+		Spec: corev1.ServiceSpec{
+			Selector: map[string]string{
+				"app.kubernetes.io/name": kubeProxyDeploymentName,
+			},
+			Ports: []corev1.ServicePort{
+				{
+					Name:       "kube-flagd-proxy",
+					Port:       kubeProxyPort,
+					TargetPort: intstr.FromInt(kubeProxyPort),
+				},
+			},
+		},
+	}
+}
+
+func newFlagdKubeProxyManifest() *appsV1.Deployment {
+	replicas := int32(1)
+	return &appsV1.Deployment{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      kubeProxyDeploymentName,
+			Namespace: currentNamespace,
+			Labels: map[string]string{
+				"app": kubeProxyDeploymentName,
+			},
+		},
+		Spec: appsV1.DeploymentSpec{
+			Replicas: &replicas,
+			Selector: &metav1.LabelSelector{
+				MatchLabels: map[string]string{
+					"app": kubeProxyDeploymentName,
+				},
+			},
+
+			Template: corev1.PodTemplateSpec{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app":                    kubeProxyDeploymentName,
+						"app.kubernetes.io/name": kubeProxyDeploymentName,
+					},
+				},
+				Spec: corev1.PodSpec{
+					ServiceAccountName: kubeProxyServiceAccountName,
+					Containers: []corev1.Container{
+						{
+							Image: fmt.Sprintf("%s:%s", kubeProxyImage, kubeProxyTag),
+							Name:  kubeProxyDeploymentName,
+							Ports: []corev1.ContainerPort{
+								{
+									Name:          "port",
+									ContainerPort: kubeProxyPort,
+								},
+							},
+							Args: []string{
+								"start",
+								"--debug",
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func (m *PodMutator) handleKubeProxy(ctx context.Context, sidecar *corev1.Container, source v1alpha1.Source) error {
+	// does the proxy exist
+	exists, ready, err := m.isKubeProxyReady(ctx)
+	if err != nil {
+		return err
+	}
+	if !exists {
+		//create, defer pod deployment
```

Comment:
For my understanding - what will happen if proxy creation continue to fail (ex:- image pull error, OOMKill) 🤔

Reply:
The annotated pod requiring a sidecar will not be injected, IMO this is acceptable, as if the app was deployed without the sidecar, or without the `kube-proxy` flags, its configuration may be wrong.

Reply:
The message `"kube-flagd-proxy is not ready, deferring pod admission"` is passed out via the deny so will be present on a describe of the replicaset, however I will add also add in an additional check to confirm whether the kube proxy deployment is not yet ready because it is in an error state and pass this out as an actual error.

Reply:
This is better and support fail fast approach. This will help administrators to deal with root causes :)